### DISCUSSION TITLE
Class based async exception handlers producing "TypeError: 'coroutine' object is not callable" errors

### DISCUSSION THREAD

**Discussion by gjdanis** [fastapi/fastapi#7411] [Questions] [answered]
### Example

```Python
import uvicorn
from fastapi import FastAPI
from starlette.requests import Request
from starlette.responses import Response, JSONResponse
from fastapi.exceptions import StarletteHTTPException, HTTPException
from fastapi.exception_handlers import http_exception_handler


class MyAysncExceptionHandler:
    async def __call__(self, request: Request, ex: Exception) -> Response:
        ex = StarletteHTTPException(detail=repr(ex), status_code=500)
        return await http_exception_handler(request, ex)


class MyNonAsyncExceptionHandler:
    def __call__(self, request: Request, ex: Exception) -> Response:
        return JSONResponse({"detail": repr(ex)}, status_code=500)


app = FastAPI()
# this works fine
app.add_exception_handler(StarletteHTTPException, MyNonAsyncExceptionHandler())


# this produces "TypeError: 'coroutine' object is not callable" message
# app.add_exception_handler(StarletteHTTPException, MyAysncExceptionHandler())

@app.get("/")
def throw_exception(message: str = "Test message"):
    raise HTTPException(detail=message, status_code=500)

if __name__ == "__main__":
    uvicorn.run(app, host="0.0.0.0", port=5000, debug=True)
```

### Description

* Open the browser and call the endpoint `/`.
* It returns JSON from the error handler
* Now comment out `app.add_exception_handler(StarletteHTTPException, MyNonAsyncExceptionHandler())`
* Use `app.add_exception_handler(StarletteHTTPException, MyAysncExceptionHandler())` instead
* You should get an internal server error hitting the endpoint again with "TypeError: 'coroutine' object is not callable" in the logs

On the surface it looks like the only difference between these exception handlers is that one is async and one is not async. Is this a bug?

### Environment

- OS: [e.g. Linux / Windows / macOS]: Windows
- FastAPI Version [e.g. 0.3.0]: 0.54.1
- Python version: Python 3.8.2

**Comment by gjdanis** [comment]
It looks like function based error handlers work fine:

```python
@app.exception_handler(StarletteHTTPException)
async def my_exception_helper(request: Request, ex: Exception):
    ex = StarletteHTTPException(detail=repr(ex), status_code=500)
    return await http_exception_handler(request, ex)
```

**Comment by tiangolo** [comment]
Hey @gjdanis , please keep the template for the issue and follow the steps.

Yes, exception handlers are expected to be functions, check the docs here: https://fastapi.tiangolo.com/tutorial/handling-errors/#install-custom-exception-handlers

**Comment by gjdanis** [comment]
Hey @tiangolo thanks for responding. I did look at the documentation and didn't see anything prohibiting class based callables. Can we improve the documentation if this is a known limitation?

**Comment by tiangolo** [comment, marked as answer]
Hmm, not sure it would really be necessary, as it's not explicitly saying that any type of callable is allowed 🤔 

I would rather keep the docs simple, so I'll pass on this suggestion. Given that I'll close the issue. But thanks for the interest! ☕ 

> Sorry for the long delay! 🙈  I wanted to personally address each issue/PR and they piled up through time, but now I'm checking each one in order.
